validation: remove redundant check on pindex

jarolrod opened the pull request:
This removes a redundant check on `pindex` being a `nullptr`. By the time we get to this step `pindex` is always a `nullptr` as the branch where it has been set would have already returned. 

Closes #19223

jarolrod committed:
validation: remove redundant check on pindex

This removes a conditional that checks if pindex is equal to nullptr.
This check is redundant because the branch where pindex is set returns at an earlier time. Additionaly, The independence of the earlier and later pindex is made clearer.

mjdietzx commented:
Concept ACK

Zero-1729 commented:
ACK

ajtowns commented:
Perhaps make it:

```c++
     BlockMap::iterator miSelf = m_block_index.find(hash);
-    CBlockIndex *pindex = nullptr;
     if (hash != chainparams.GetConsensus().hashGenesisBlock) {
         if (miSelf != m_block_index.end()) {
             // Block header is already known.
-            pindex = miSelf->second;
+            CBlockIndex *pindex = miSelf->second;
             if (ppindex)
                 *ppindex = pindex;
             if (pindex->nStatus & BLOCK_FAILED_MASK) {
@@ -3621,7 +3620,7 @@ bool BlockManager::AcceptBlockHeader(const CBlockHeader& block, BlockValidationS
             }
         }
     }
-    pindex = AddToBlockIndex(block);
+    CBlockIndex *pindex = AddToBlockIndex(block);
 ```
 
 so it's clear that the earlier `pindex` and the later `pindex` are independent?

jarolrod commented:
updated 1574879a7b06e570c9943de02e6efcc74817d44a -> 8be1a34f56ab8cb76b2d29b081c77c052bbad191

addressed @ajtowns [comment](https://github.com/bitcoin/bitcoin/pull/20868#issuecomment-755891663). Thanks for the tip!

promag commented on src/validation.cpp:
```diff
@@ -3621,8 +3620,7 @@ bool BlockManager::AcceptBlockHeader(const CBlockHeader& block, BlockValidationS
     }
-    if (pindex == nullptr)
-        pindex = AddToBlockIndex(block);
+    CBlockIndex *pindex = AddToBlockIndex(block);
```
nit, `CBlockIndex* pindex = ...`, same in L3554.

promag reviewed (commented):
Code review ACK 8be1a34f56ab8cb76b2d29b081c77c052bbad191.

theStack reviewed (approved):
Code review ACK 8be1a34f56ab8cb76b2d29b081c77c052bbad191 🥃

theStack reviewed (approved):
re-ACK c943282b5e6312537f885c811d43120ce2f5b766

jarolrod commented:
updated 8be1a34f56ab8cb76b2d29b081c77c052bbad191 -> c943282b5e6312537f885c811d43120ce2f5b766

Moved placement of pointer from `name` to `type` based on @promag [suggestion](https://github.com/bitcoin/bitcoin/pull/20868#discussion_r554572500)

Zero-1729 commented:
re-ACK c943282

ajtowns commented:
ACK c943282b5e6312537f885c811d43120ce2f5b766 - code review only

decryp2kanon commented:
Concept ACK

MarcoFalke commented:
review ACK c943282b5e6312537f885c811d43120ce2f5b766 📨

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK c943282b5e6312537f885c811d43120ce2f5b766 📨
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhzXQwAyGILsbFtO9WYGBALSWon9GjpoziL7M+3LqQBj7D21YVmRmCZAL4FV8+8
++ebIBccAaI84vRn+TAjY74UYHv55EsxnwbZQzY44KivRc6kWUlf1AubA16WrW0z
jl3uP4qImGpy2nflWUjGsKkVtWg+Y/fAmgM0v2HiEGRsz+GmyidMpeIqFSjd07A4
uyavf+dC3BK8q9kM7EYhH7lZxlrbti9cdtM5f8gfccYWJg33tpBrEJAwJ0EXKjEh
6vYTEf7XYXlSX3FZQZ9q2fyyUecvfo05+G15anfGSGqzKnE5u4n438VamUuQtM3b
yQyMddEnmI0iodXOIRGtZhuxwFh7RDdyPUdG//fPqhHrvN4/7LG3vP1aBAO0MRZU
kT+pQWxvHHgCun7hgu4SCfdAMP2vFZ0aPTatj98KQ/t4wMJcboEVAL6l7c1aeOgU
GM/S9HoT3mjiV8VZTZ4ykC1HcgZFHRaSWiBRseJ8GmAg7DLKkTu9d/LFkX8VqdSY
pxX7K5l9
=o4SE
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `76e16b7b369644db399a29534fee3202bd7da8a1fc17996983ad5d72b5c4ff4a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010876e16b7b369644db399a29534fee3202bd7da8a1fc17996983ad5d72b5c4ff4af010e8feeb18b01cb93d2865b887c066368808fff010227ed51591ab83a91b39760a645c033808f1046017cff0f00879b5c966d6d4e1db0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01057e43e9a2a97f6d0a831959d14d996b008f1046017cff0f008c3a27e6464c925010083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107c29e6a950adde66f836de0895f0484508f1046017c21ff008236e0bb201ca7cee0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010de15b6a8e5c973712200ec926a235c7808f1046017cff0f008f54923e5b89e8e270083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
